// RUN_PIPELINE_TILL: FRONTEND
// LANGUAGE: +ContextParameters

<!UNSUPPORTED_FEATURE!><!CONTEXT_RECEIVERS_DEPRECATED!>context<!>(s: String)<!>
class C {
    <!UNSUPPORTED_FEATURE!><!CONTEXT_RECEIVERS_DEPRECATED!>context<!>(s: String)<!>
    constructor() {}

    <!UNSUPPORTED_FEATURE!><!CONTEXT_RECEIVERS_DEPRECATED!>context<!>(s: String)<!>
    fun f(){}

    <!UNSUPPORTED_FEATURE!><!CONTEXT_RECEIVERS_DEPRECATED!>context<!>(_: String)<!>
    val p: String get() = ""
}

<!UNSUPPORTED_FEATURE!><!CONTEXT_RECEIVERS_DEPRECATED!>context<!>(s: String)<!>
fun f(){}

<!UNSUPPORTED_FEATURE!><!CONTEXT_RECEIVERS_DEPRECATED!>context<!>(_: String)<!>
val p: String get() = ""